### ISSUE TITLE
Websockets connection is not robust

### ISSUE THREAD

**Issue by jochenkrattenmacher** [GenieFramework/Genie.jl#613]
**Describe the bug**
It happens to me quite often that I see the banner which says "Can not reach the server - please reload the page", mostly when accessing the hosted page from other devices via a local wifi (never hosted on the Internet, and accessing the page from the hosting device comes with a rather robust connection). Side note: If I have multiple tabs open in the same browser, they often seem to lose connection at the same time.

**Error stacktrace**
There are no errors in the REPL. The console of the client says 
`The connection to ws://192.168.178.37:8000/ was interrupted while the page was loading`. Below that message one sees an error `undefined`, stacktrace mentions lines 39 (`f(event);`) and 142 (`console.error(event.data);`) in channels.js (FYI, I here changed the line numbers such to correspond to the respective lines in https://github.com/GenieFramework/Genie.jl/blob/master/assets/js/channels.js) 

**To reproduce**
Most reliably the connection loss seems to happen on my phone (with Chrome. Firefox connections on my phone seem a bit more robust). I am hosting on my laptop within my local wifi, and the below code is the most minimal example I tried:

```julia
using Stipple, StippleUI

@reactive! mutable struct Name <: ReactiveModel
end

function ui(model)
  [
    page(
      model,
    ),
  ]
end

route("/") do
  init(Name) |> ui |> html
end

up(8000, "192.168.178.37")
```
(it is sufficient to simply load the page, twirl one's thumbs for 10 minutes or so, and then look at the page again)

**Additional context**
This happens to me on multiple machines, Linux, MacOS, Windows, from Julia 1.7.2 to 1.8.2, and Chrome and Firefox on the client-side (didn't try other browsers)

### Partial fix
Changing the function `sendMessageTo` in channels.js to the below has made the connection from the client to the server much more robust:

```julia
function sendMessageTo(channel, message, payload = {}) {
  if (Genie.WebChannels.socket.readyState === 1) {
    Genie.WebChannels.socket.send(JSON.stringify({
      'channel': channel,
      'message': message,
      'payload': payload
    }));
  } else if (Genie.WebChannels.socket.readyState > 1) {
    Genie.WebChannels.socket = newSocketConnection();
    subscribe();
    if (Genie.WebChannels.socket.readyState === 1) {
      Genie.WebChannels.socket.send(JSON.stringify({
        'channel': channel,
        'message': message,
        'payload': payload
      }));
    } else {
      console.warn('WebSocket is not open');
      displayAlert();
    }
  }
}
```

(to further improve upon the above, it might be better to [wait for the websocket connection to be open before sending the message](https://dev.to/ndrbrt/wait-for-the-websocket-connection-to-be-open-before-sending-a-message-1h12))

The banner appears much less frequently now, and even if it does, I typically, if not always, am still able to send data from the client to the server (e.g., ticking a checkbox would change the respective field in the backend). However, in that state, the client won't receive messages from the server anymore (i.e., changing a field in the backend won't change the checkbox status).

I might try to look at the server-to-client communication this week, but it seems a bit more tricky to me, and I'd be grateful for pointers.

**Comment by jochenkrattenmacher** [comment]
So I think I have a full fix now (page keeps connection to backend), but I'll test it more tomorrow and Friday - if something turns up, I'll write another post.

I solely edited channels.js for this fix. Here's the corresponding commit (FYI: the changes I mentioned in the first post are not needed anymore with this fix): https://github.com/jochenkrattenmacher/Genie.jl/commit/c85809b15294398af07fb304490fddd021ed43e5

basically, now the websocket is re-initialized and resubscribed to upon connection loss. The re-initialization includes setting up all the socket-associated listeners. **As far as I can see, this also fixes the inbuilt Genie reconnect functionality, which thus-far probably didn't work correctly because the listeners were not regenerated? I at least once had encountered a state where the client could send data to the host, but not vice versa...**

The @info calls in WebChannels.jl are helpful for understanding what is happening. If I connect to the host with Chrome on my phone, I will see something like the following in the REPL (if I wait long enough):

```
[ Info: Subscribed: 1972089144317918958
[ Info: Subscribed: 16730352834621593616
[ Info: Subscribed: 16730352834621593616
```

The "trial" argument passed to subscribe() simply keeps track of how many times subscribe was called already to avoid battery drainage in case devices are offline or the server is down. If trial reaches 4, the "Can not reach the server - please reload the page" banner is shown

**Comment by essenciary** [comment]
@jochenkrattenmacher excellent updates, thanks so much! Can you please make a PR which the fix? Happy to test and merge.

**Comment by jochenkrattenmacher** [comment]
thanks, created a PR :)

In the latest commit, I implemented above-mentioned method for [waiting for the websocket connection to be open before sending the message](https://dev.to/ndrbrt/wait-for-the-websocket-connection-to-be-open-before-sending-a-message-1h12). I am not sure whether using eval is really necessary, but it works this way (the thinking goes that the socket is a different one if one has resubscribed in the meantime).

**Comment by jochenkrattenmacher** [comment]
I did some more testing today with multiple devices and browser tabs, and I indeed found it to be much more stable now. I think at some point my server was under too much load and thus it took a while for the messages from the clients to be processed, but I suppose I could improve on that by not using Genie for serving static content. Once, a client lost connection completely and couldn't send a message, so there is still some potential for further improvements I suppose.

Side notes: 
- I suppose the setting "webchannels_server_gone_alert_timeout" now is somewhat unnecessary. It might be better to simply show the alert immediately, or introduce a setting with which users can customize the alert (perhaps the full function?), also allowing for not showing/doing anything at all. I would use such a setting for instructing the page to reload immediately without showing anything (but I suppose that's an edge case).
- It might be good to remove the alert (and the "#wsconnectionalert" in the URL) once the  client has reconnected to the server. Although I suppose one needs to tread carefully here because the client might be out of sync with the server in such cases, and if the banner isn't there, the user might not know (though that also might not necessarily be clear to users currently, given that the banner doesn't inform about that possibility). Perhaps the user could be disallowed from making any changes to the vue model while the connection is lost (e.g., checkboxes would be disabled)
